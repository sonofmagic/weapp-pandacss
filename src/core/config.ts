import { loadConfigFile } from '@pandacss/config'
import { defu } from 'defu'
import { loadConfig, UserInputConfig, createDefineConfig } from 'c12'
import { getCreateContextDefaults, getPostcssPluginDefaults } from '../defaults'
import { PandacssConfigFileOptions, UserConfig } from '@/types'
export function getPandacssConfig(
  options?: Partial<PandacssConfigFileOptions>
) {
  const opt = defu<PandacssConfigFileOptions, PandacssConfigFileOptions[]>(
    options,
    {
      cwd: process.cwd()
    }
  )

  return loadConfigFile(opt)
}

export function getConfig(options?: Pick<UserInputConfig, 'cwd'>) {
  return loadConfig<UserConfig>({
    name: 'weapp-pandacss', // `${name}.config` //
    rcFile: false,
    globalRc: false,
    cwd: options?.cwd,
    defaults: {
      context: getCreateContextDefaults(),
      postcss: getPostcssPluginDefaults()
    }
  })
}
export const defineConfig = createDefineConfig<UserConfig>()
// export function defineConfig(config: UserConfig) {
//   return config
// }
