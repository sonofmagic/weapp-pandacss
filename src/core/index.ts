export * from './context'
export * from './codegen'
export * from './config'
export * from './patch'
export * from './logger'
