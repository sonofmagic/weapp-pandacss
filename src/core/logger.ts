import { bold, green, cyan } from 'colorette'

export const tick = bold(green('✔️'))

export const quote = (...str: string[]) => cyan(`\`${str.join('')}\``)
