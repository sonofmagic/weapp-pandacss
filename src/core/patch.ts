import fs from 'node:fs/promises'
import * as t from '@babel/types'

import { generate, parse, traverse } from '@/babel'

export function inject(content: string) {
  const root = parse(content, {
    sourceType: 'unambiguous'
  })

  traverse(root, {
    Program: {
      enter(p) {
        // import { escape } from './weapp-panda/index.mjs'
        const node = p.node
        const importNodes = node.body.filter(
          (x) => x.type === 'ImportDeclaration'
        ) as t.ImportDeclaration[]
        const myImport = importNodes.find(
          (x) => x.source.value === './weapp-panda/index.mjs'
        )
        if (!myImport) {
          node.body.unshift(
            t.importDeclaration(
              [
                t.importSpecifier(
                  t.identifier('escape'),
                  t.identifier('escape')
                )
              ],
              t.stringLiteral('./weapp-panda/index.mjs')
            )
          )
        }
      }
    },
    FunctionDeclaration: {
      enter(p) {
        const node = p.node
        if (node.id?.name === 'createCss') {
          const returnFn = node.body.body.find(
            (x) => x.type === 'ReturnStatement'
          ) as t.ReturnStatement | undefined
          if (
            returnFn &&
            returnFn.argument?.type === 'ArrowFunctionExpression'
          ) {
            const arrowFn = returnFn.argument
            if (arrowFn.body.type === 'BlockStatement') {
              const innerReturn = arrowFn.body.body.find(
                (x) => x.type === 'ReturnStatement'
              ) as t.ReturnStatement | undefined
              if (innerReturn) {
                // import { escape } from '@weapp-core/escape'
                const originNode = innerReturn.argument as t.CallExpression
                if (
                  originNode.callee.type === 'Identifier' &&
                  originNode.callee.name === 'escape'
                ) {
                  // has patched
                  return
                }

                innerReturn.argument = t.callExpression(
                  t.identifier('escape'),
                  [originNode]
                )
              }
            }
          }
        }
      }
    }
  })

  return generate(root)
}

export async function patch(src: string, dest?: string) {
  const content = await fs.readFile(src, 'utf8')
  const { code } = inject(content)
  await fs.writeFile(dest ?? src, code, 'utf8')
}
