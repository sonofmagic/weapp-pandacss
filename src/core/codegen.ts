import fs from 'node:fs/promises'
import path from 'node:path'
import { dedent } from './logger'
import { ICreateContextOptions } from '@/types'
export async function ensureDir(p: string) {
  try {
    await fs.access(p)
  } catch {
    await fs.mkdir(p, {
      recursive: true
    })
  }
}

export function getWeappCoreEscapeDir() {
  return path.dirname(require.resolve('@weapp-core/escape'))
}

export function getPandaVersion() {
  return require(require.resolve('@pandacss/dev/package.json')).version
}

// dirName: string = 'weapp-panda'
export async function copyEscape(destDir: string) {
  const result: string[] = []
  const srcDir = getWeappCoreEscapeDir()
  const filesnames = await fs.readdir(srcDir)
  await ensureDir(destDir)

  for (const filesname of filesnames) {
    const src = path.resolve(srcDir, filesname)
    const stats = await fs.stat(src)
    if (stats.isFile()) {
      const dest = path.resolve(destDir, filesname)
      result.push(dest)
      await fs.copyFile(src, dest)
    }
  }
  return result
}

export async function generateEscapeWrapper(
  destDir: string,
  options: ICreateContextOptions
) {
  await ensureDir(destDir)
  const code = dedent`
  import { escape as _escape } from './lib/index.mjs'
  function escape(selectors) {
    if(${options.escapePredicate}){
      return _escape(selectors)
    }
    return selectors
  }
  export { escape }
  `
  await fs.writeFile(path.resolve(destDir, 'index.mjs'), code, 'utf8')
  await fs.writeFile(
    path.resolve(destDir, 'index.d.ts'),
    dedent`
 declare function escape(selectors: string): string;`,
    'utf8'
  )
}
