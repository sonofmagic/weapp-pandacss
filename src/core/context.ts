import { resolve, dirname } from 'node:path'
import fs from 'node:fs/promises'
import { existsSync } from 'node:fs'
import { defu } from 'defu'
import { getPandacssConfig } from './config'
import { copyEscape } from './codegen'
import { patch } from './patch'
import { tick, quote, dedent } from './logger'
import { ICreateContextOptions } from '@/types'
import { getCreateContextDefaults } from '@/defaults'

export async function createContext(options?: ICreateContextOptions) {
  const opt = defu(options, getCreateContextDefaults())

  const pandaConfig = await getPandacssConfig(opt.pandaConfig)

  async function codegen() {
    const words: string[] = []
    const outdir = pandaConfig.config.outdir
    const weappPandaDir = resolve(outdir, 'weapp-panda')
    const patchHelpersPath = resolve(outdir, 'helpers.mjs')
    await copyEscape(weappPandaDir)
    words.push(dedent`
    ${tick} ${quote(outdir, '/weapp-panda')}: the core escape function for weapp
    `)
    if (existsSync(patchHelpersPath)) {
      await fs.copyFile(
        patchHelpersPath,
        resolve(dirname(patchHelpersPath), '_helpers.backup.mjs')
      )
    }

    await patch(patchHelpersPath)
    words.push(dedent`
    ${tick} ${quote(
      outdir,
      '/helpers.mjs'
    )}: inject escape function into helpers
    `)
    console.log(words.filter(Boolean).join('\n'))
  }

  async function rollback() {
    const outdir = pandaConfig.config.outdir
    const patchHelpersBackupPath = resolve(outdir, '_helpers.backup.mjs')
    if (existsSync(patchHelpersBackupPath)) {
      await fs.copyFile(
        patchHelpersBackupPath,
        resolve(dirname(patchHelpersBackupPath), 'helpers.mjs')
      )
    }
  }
  return {
    pandaConfig,
    codegen,
    rollback
  }
}
