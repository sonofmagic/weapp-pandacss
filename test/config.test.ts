import { appRoot, taroAppRoot } from './util'
import { getPandacssConfig, getConfig } from '@/core/config'
describe('config', () => {
  it('get fixtures app config', async () => {
    const config = await getPandacssConfig({
      cwd: appRoot
    })
    expect(config).toBeDefined()
    expect(config.config.outdir === 'src/styled-system').toBe(true)
  })

  it('get fixtures taroApp config', async () => {
    const config = await getPandacssConfig({
      cwd: taroAppRoot
    })
    expect(config).toBeDefined()
    expect(config.config.outdir === 'styled-system').toBe(true)
  })

  it('get default config', async () => {
    const { config, ...rest } = await getConfig()
    console.log(rest)
    expect(config).toMatchSnapshot()
  })
})
