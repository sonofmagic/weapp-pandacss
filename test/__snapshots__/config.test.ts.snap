// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`config > get default config 1`] = `
{
  "context": {
    "escapePredicate": "true",
    "log": false,
  },
  "postcss": {
    "cascadeLayersPluginOptions": {},
    "disabled": false,
    "isPseudoClassPluginOptions": {},
    "removeNegationPseudoClass": true,
    "selectorReplacement": {
      "cascadeLayers": "n",
      "root": "page",
      "universal": [
        "view",
        "text",
      ],
    },
  },
}
`;

exports[`config > get default config from 0.default dir 1`] = `
{
  "context": {
    "escapePredicate": "true",
    "log": false,
    "pandaConfig": {
      "cwd": "/Users/icebreaker/Documents/GitHub/weapp-pandacss",
    },
  },
  "escapePredicate": [Function],
  "postcss": {
    "cascadeLayersPluginOptions": {},
    "disabled": true,
    "isPseudoClassPluginOptions": {},
    "removeNegationPseudoClass": true,
    "selectorReplacement": {
      "cascadeLayers": "n",
      "root": [],
      "universal": [
        "view",
        "text",
      ],
    },
  },
}
`;
